# PZM-Tools

This repository houses scripts for classifying post-zygotic indel 
and SNV variants as either true or false positives. 
The classification is based on a random forest model trained using 
molecularly validated variants to distinguish real post-zygotic 
mutations from others.

The script takes as input a VCF file generated by 
[Mutect2](https://gatk.broadinstitute.org/hc/en-us/articles/360037593851-Mutect2),
and inputs should contain the same attributes in the `INFO`
and `FORMAT` fields as Mutect2 output.
[This directory](https://github.com/talkowski-lab/pzm/tree/main/src/test_data)
contains example VCF files and their corresponding index.

This script is available as a Python package, Docker container,
and WDL workflow. You can run it locally, via command line, 
in Docker, or on the cloud with Cromwell or Terra.
Please refer to [this README](https://github.com/talkowski-lab/pzm/blob/main/src/README.md)
for details.
